compiler/src/iree/compiler/Codegen/LLVMGPU/AMDGPUChainedMatmulPass.cpp

```diff
@@ -85,22 +136,65 @@ struct AMDGPUPrepareForChainedMatmulPass
     return newIndexingMaps == contractOp.getIndexingMapsArray();
   }
 
+  /// Returns a vector.contract operation that this value was transitively
+  /// produced from.
+  ///
+  /// A chained matmul is one where the lhs of the candidate matrix
+  /// is a result of another matmul (a matmul lies in the backward slice of lhs
+  /// of the first matmul).
+  FailureOr<vector::ContractionOp>
+  getTransitiveMatmulParent(vector::ContractionOp contractOp) const {
+    SetVector<Operation *> backwardSlice;
+    getBackwardSlice(contractOp.getLhs(), &backwardSlice);
+    vector::ContractionOp result;
+    for (Operation *sliceOp : backwardSlice) {
+      auto chainParent = dyn_cast<vector::ContractionOp>(sliceOp);
+      if (!chainParent) {
+        continue;
+      }
+
+      // For now, we only support transpose invariant matmuls. This is because
+      // transposing the inputs may have a non-trivial cost which we need
+      // to think about.
+      if (!isOperandSwapInvariant(chainParent)) {
+        continue;
+      }
+
+      // If we have multiple matmul parents, we fail.
+      if (result) {
+        return failure();
+      }
+
+      result = chainParent;
+    }
+
+    if (result) {
+      return result;
+    }
+
+    return failure();
+  }
+
   void runOnOperation() override {
     auto funcOp = getOperation();
-    SmallVector<vector::ContractionOp> chainedMatmuls;
+    SmallVector<vector::ContractionOp> matmulCandidates;
     funcOp.walk([&](vector::ContractionOp contractOp) {
-      if (!isCompatibleIndexingMap(contractOp, funcOp.getContext()))
-        return WalkResult::advance();
-      chainedMatmuls.push_back(contractOp);
-      return WalkResult::advance();
+      matmulCandidates.push_back(contractOp);
     });
-    if (chainedMatmuls.size() != 2)
-      return;
-    if (!isChainedMatmul(chainedMatmuls))
-      return;
+
```

Comment:
Shouldn't we have a constraint of the number of matmuls? Would this work if we had 3 chained matmuls for example?

Reply:
Ideally, it would be. Our chained matmul needs to be checking the layout produced and expected by the chain. I added a TODO for this.